keps/sig-api-machinery/00xx-admission-webhooks-to-ga.md

```diff
@@ -0,0 +1,645 @@
+---
+kep-number: 36
+title: Graduate Admission Webhooks to GA
+authors:
+  - "@mbohlool"
+owning-sig: sig-api-machinery
+reviewers:
+  - "@liggitt"
+  - "@deads2k"
+approvers:
+  - "@liggitt"
+  - "@deads2k"
+editor: TBD
+creation-date: 2019-01-27
+last-updated: 2019-01-29
+status: provisional
+see-also:
+  - [Admission Control Webhook Beta Design Doc](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/api-machinery/admission-control-webhooks.md)
+---
+
+# Admission Webhooks to GA
+
+## Table of Contents
+
+* [Summary](#summary)
+* [Motivation](#motivation)
+  * [Goals](#goals)
+  * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+  * [Object selector](#object-selector)
+  * [Scope](#scope)
+  * [http for localhost](#http-for-localhost)
+  * [timeout configuration](#timeout-configuration)
+  * [Port configuration](#port-configuration)
+  * [Plumbing existing objects to delete admission](#plumbing-existing-objects-to-delete-admission)
+  * [Mutating Plugin ordering](#mutating-plugin-ordering)
+  * [Passing {Operation}Option to Webhook](#passing-operationoption-to-webhook)
+  * [AdmissionReview v1](#admissionreview-v1)
+* [V1 API](#v1-api)
+* [V1beta1 changes](#v1beta1-changes)
+* [Validations](#validations)
+* [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+* [Post-GA tasks](#post-ga-tasks)
+* [Implementation History](#implementation-history)
+
+## Summary
+
+Admission webhook is a way to extend kubernetes by putting hook on object creation/modification/deletion. Admission webhooks can mutate or validate the object. This feature has been Beta since Kubernetes 1.9. This document outline required steps to graduate it to GA.
+
+## Motivation
+
+Admission webhooks are currently widely used for extending kubernetes but has been in beta for 3 releases. Current set of feature requests and bug reports on the feature shows it is close to be stable and motivation of this KEP is to address those feature requests and bug reports to move this feature toward general availability (GA).
+
+### Goals
+
+Based on the user feedback, These are the planned changes to current feature to graduate it to GA:
+
+* Object Selector: Add a general selector for objects that a webhook can target.
+* Scope: Define the scope of the targeted objects, e.g. cluster vs. namespaced.
+* http for localhost: relax https constraint for localhost.
+* timeout configuration: Add a configuration field to shorten the timeout of the webhook call.
+* port configuration: Add a configuration field to change the service port from the default 443.
+* AdmissionReview version selection: Add a API field to set the AdmissionReview versions a webhook can understand.
+* Pass existing objects to delete admission webhooks
+* re-run mutating plugins if any webhook changed object to fix the plugin ordering problem
+* pass OperationOption (such as CreateOption/DeleteOption) to the webhook
```

Comment:
Add: Webhook Configuration should not require users/admins to understand every possible way a user can write an object (e.g., you must hook both apps/deployment and extensions/deployment).

Reply:
I remember I discussed this early on with @liggitt and he recommended we do not make this item a requirement for GA. Let's add this item in a follow up PR if we wanted to do it.

Reply:
We [discussed this on Wednesday](https://docs.google.com/document/d/1x9RNaaysyO0gXHIr1y50QFbiL1x8OWnk2v3XnrdkT5Y/edit#bookmark=id.w120b3estxno), and I think it's worth attempting for GA. All the pros are for webhook authors and could significantly change the way you'd write/register a webhook (write admission for one version of a resource, don't worry about new versions or groups), and all the cons are for the apiserver authors (performance cost of converting up to *2n* times in order to dispatch to *n* webhooks, some of those conversions possibly being done by CRD conversion webhooks, etc). I'm ok adding it as a planned item, but there are some interesting edges to work through before this bit would be implementable.